### PR TITLE
fix #1013, track task checkpoint time at ClockService

### LINKED ISSUE #1013: track task checkpoint time at ClockService

### PR DESCRIPTION
**Pull request by manuzhang**
[Review on Reviewable](https://reviewable.io/reviews/intel-hadoop/gearpump/1014)

### CODE REVIEW

**Comment by manuzhang** [comment]
some UTs failed. I'm fixing them now.

**Comment by manuzhang** [comment]
remaining test failures are unrelated

File: streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskManager.scala [hunk 1]
```diff
@@ -121,7 +121,7 @@ private[appmaster] class TaskManager(
 
     case Event(RegisterTask(taskId, executorId, host), state@ TaskRegistrationState(register)) =>
       val client = sender
-      getMinClock.map(client ! StartClock(_))
+      getLatestStartClock.map(client ! StartClock(_))
```
Comment (clockfly) [inline streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskManager.scala]: How can we make the fault-tolerance level configurable?
at-least once
exactly-once
at mostly once

Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskManager.scala]: at API level

at most once - by default
at least once - TimeReplayableSource
exact once - TimeReplayableSource + PersistentTask

Comment (clockfly) [inline streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskManager.scala]: For at least once, should we replay from LatestStartClock or LatestMinClock?

Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskManager.scala]: I think we should replay from `LatestStartClock`. 
```scala
private def latestStartClock: TimeStamp = {
   minCheckpointTime.getOrElse(minClock)
}
```
when it's at least once, `LatestStartClock` returns the value of `LatestMinClock` so it covers both at least once and exact once for replay.

**Comment by manuzhang** [comment]
Any further comments ?

**Comment by clockfly** [comment]
@manu

The code base like TaskManager, ClockService in this PR is completely rewritten in DynamicDAG.
Let's hold the PR for some while so that later we can port the change to new code base.

**Comment by manuzhang** [comment]
sure. I can rebase on new codebase later.

**Comment by manuzhang** [comment]
sure. I can rebase on new codebase later.

**Comment by manuzhang** [comment]
closed to rebase on latest trunk and not block package name refactoring

**Pull request closed without merging**